src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size = 0;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void grow() {
+        int newCapacity = (int) (elements.length * 1.5);
+        elements = Arrays.copyOf(elements, newCapacity);
+    }
+
+    private void growTo(int minCapacity) {
+        int newCapacity = elements.length;
+        while (newCapacity < minCapacity) {
+            newCapacity = (int) (newCapacity * 1.5);
+        }
+        elements = Arrays.copyOf(elements, newCapacity);
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            grow();
+        }
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
```

Comment:
The index validation logic is repeated in multiple methods. Consider extracting this logic into a separate private method to adhere to the DRY principle.